### PR TITLE
Support portable.exe apps and better error reporting

### PR DESCRIPTION
**Pull request by pakrym**
I'm going to fix/add more tests.

### CODE REVIEW

File: src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp [hunk 1]
```diff
@@ -3,344 +3,198 @@
...
-    const fs::path applicationPhysicalPath = pcwzApplicationPhysicalPath;
-    fs::path processPath = Environment::ExpandEnvironmentVariables(pcwzProcessPath);
-    std::wstring arguments = Environment::ExpandEnvironmentVariables(pcwzArguments);
+    LOG_INFOF("Expanded hostfxr parameters for application: '%S' arguments: '%S'",
+        expandedProcessPath.c_str(),
+        expandedApplicationArguments.c_str());
 
-    // Check if the absolute path is to dotnet or not.
-    if (IsDotnetExecutable(processPath))
+    LOG_INFOF("Known dotnet.exe location: '%S'", dotnetExePath.c_str());
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp]: We do, it's the only thing that we pass across shim-rh boundary.

File: src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp [hunk 2]
```diff
@@ -3,344 +3,198 @@
...
 
-        processPath = fullProcessPath.value();
+    // Check if the absolute path is to dotnet or not.
+    if (IsDotnetExecutable(expandedProcessPath))
+    {
+        LOG_INFOF("Process path '%S' is dotnet, treating application as portable", expandedProcessPath.c_str());
 
-        auto hostFxrPath = GetAbsolutePathToHostFxr(processPath);
-        if (!hostFxrPath.has_value())
+        if (dotnetExePath.empty())
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp]: When we run this method in shim `dotnetExePath` would be empty

**Comment by jkotalik** [comment]
d:\j\workspace\windows-confi---dfc477d1\src\aspnetcoremodulev2\commonlib\handlewrapper.h(26): error C2131: expression did not evaluate to a constant (compiling source file FileOutputManager.cpp) [D:\j\workspace\windows-Confi---dfc477d1\src\AspNetCoreModuleV2\CommonLib\CommonLib.vcxproj] [C:\Users\runner\.dotnet\buildtools\korebuild\2.2.0-preview1-20180816.6\KoreBuild.proj]

File: src/AspNetCoreModuleV2/CommonLib/hostfxroptions.h [hunk 3]
```diff
@@ -5,41 +5,45 @@
...
-    }
-
-    DWORD
-    GetArgc() const
+    HOSTFXR_OPTIONS(
+        std::filesystem::path dotnetExeLocation,
+        std::filesystem::path hostFxrLocation,
+        std::vector<std::wstring> arguments
+        )
+    :   m_dotnetExeLocation(std::move(dotnetExeLocation)),
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/hostfxroptions.h]: https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#pragma once
+
+#include <string>
+
+[[nodiscard]]
+bool ends_with(const std::wstring &source, const std::wstring &suffix, bool ignoreCase = false);
+
+template<typename ... Args>
+[[nodiscard]]
+std::wstring format(const std::wstring& format, Args ... args)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/StringHelpers.h]: nit: either move this to StringHelpers.cpp or consolidate into one file.

File: src/AspNetCoreModuleV2/CommonLib/aspnetcore_msg.mc [hunk 5]
```diff
@@ -38,12 +38,6 @@ Language=English
 %1
 .
 
-Messageid=1002
-SymbolicName=ASPNETCORE_EVENT_PROCESS_CRASH
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/aspnetcore_msg.mc]: Interesting how ASPNETCORE_EVENT_PROCESS_CRASH isn't used by out of process.

File: src/AspNetCoreModuleV2/CommonLib/StringHelpers.h (hunk 4 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/StringHelpers.h]: `template<typename ... Args>` method can't be moved to cpp unfortunately.

File: src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp [hunk 6]
```diff
@@ -3,344 +3,197 @@
...
-                hr);
-            return hr;
+            throw StartupParametersResolutionException(format(L"Executable was not found at '%s'", executablePath.c_str()));
         }
     }
+}
 
-    return S_OK;
+BOOL
+HOSTFXR_UTILITY::IsDotnetExecutable(const std::filesystem::path & dotnetPath)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp]: Nit: Consider dropping this this function entirely and just calling ends_with. It's only used once.

**Pull request merged**